rs/sns/testing/src/utils.rs

```diff
@@ -0,0 +1,221 @@
+    #[error("Failed to get SNS subnet IDs: {0}")]
+    FailedToGetSnsSubnetIds(String),
+    #[error("SNS WASM doesn't have any upgrade steps")]
+    MissingSnsWasmUpgradeSteps,
+    #[error("Failed to get SNS WASM upgrade steps: {0}")]
+    FailedToGetSnsWasmUpgradeSteps(String),
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum SnsTestingCanisterValidationError {
+    #[error("Canister {0} is not installed")]
+    CanisterNotInstalled(CanisterId),
+    #[error("Canister {0} is not controlled by the NNS Root canister")]
+    CanisterNotControlledByNnsRoot(CanisterId),
+    #[error("Failed to get canister '{0}' info: {1}")]
+    FailedToGetCanisterInfo(CanisterId, String),
+}
+
+// Function that validates the provided IC network setup.
+// It ensures that all required NNS canisters exist and are installed
+// and the network has at least one SNS subnet.
+pub async fn validate_network<C: CallCanisters>(
+    agent: &C,
+) -> Vec<SnsTestingNetworkValidationError> {
+    let canisters_installed = join_all(
+        ALL_NNS_CANISTER_IDS
+            .iter()
+            .map(|canister_id| async { check_canister_installed(agent, canister_id).await }),
+    )
+    .await;
+    let (_, validation_errors): (Vec<_>, Vec<_>) = ALL_NNS_CANISTER_IDS
+        .iter()
+        .zip(canisters_installed)
+        .map(|(canister_id, installed)| {
+            if !installed {
+                Err(SnsTestingNetworkValidationError::MissingNnsCanister(
+                    canister_id_to_nns_canister_name(**canister_id),
+                ))
+            } else {
+                Ok(())
+            }
+        })
+        .partition(Result::is_ok);
+    let mut validation_errors: Vec<SnsTestingNetworkValidationError> = validation_errors
+        .into_iter()
+        .map(Result::unwrap_err)
+        .collect();
+    match get_sns_subnet_ids(agent).await {
+        Ok(response) => {
+            if response.sns_subnet_ids.is_empty() {
+                validation_errors.push(SnsTestingNetworkValidationError::MissingSnsSubnets);
+            }
+        }
+        Err(err) => {
+            validation_errors.push(SnsTestingNetworkValidationError::FailedToGetSnsSubnetIds(
+                err.to_string(),
+            ));
+        }
+    }
+
+    match list_upgrade_steps(agent, None, None, 0).await {
+        Ok(response) => {
+            if response.steps.is_empty() {
+                validation_errors
+                    .push(SnsTestingNetworkValidationError::MissingSnsWasmUpgradeSteps);
+            }
+        }
+        Err(err) => {
+            validation_errors.push(
+                SnsTestingNetworkValidationError::FailedToGetSnsWasmUpgradeSteps(err.to_string()),
+            );
+        }
+    }
+    validation_errors
+}
+
+// Function that validates that provided canister is suitable to be used in SNS.
+// It ensures that canister exists, is installed and controlled by the NNS Root canister.
+pub async fn validate_target_canister<C: CallCanisters>(
```

Comment:
Part of the job of this function is to validate that the following command from `dfx sns add-nns-root` did the right thing: https://github.com/dfinity/ic/blob/5ed9bb96f3891891c1e36e299e53a96025c17d01/rs/sns/cli/src/prepare_canisters.rs#L34

As a low-prio request, could we implement a better version of `add-sns-root` that doesn't just invoke DFX? Then I would suggest usign this function to test it.